src/Sema/LowerZon.zig

```diff
@@ -53,13 +53,174 @@ pub fn run(
+        .float_literal => |val| {
+            const result = try pt.floatValue(.fromInterned(.comptime_float_type), val);
+            return result.toIntern();
+        },
+        .char_literal => |val| return pt.intern(.{ .int = .{
+            .ty = .comptime_int_type,
+            .storage = .{ .i64 = val },
+        } }),
+        .enum_literal => |val| return pt.intern(.{
+            .enum_literal = try ip.getOrPutString(
+                gpa,
+                pt.tid,
+                val.get(self.file.zoir.?),
+                .no_embedded_nulls,
+            ),
+        }),
+        .string_literal => |val| {
+            const ip_str = try ip.getOrPutString(gpa, pt.tid, val, .maybe_embedded_nulls);
+            const result = try self.sema.addStrLit(ip_str, val.len);
+            return result.toInterned().?;
+        },
+        .empty_literal => {
+            const ty = try ip.getTupleType(
+                gpa,
+                pt.tid,
+                .{
+                    .types = &.{},
+                    .values = &.{},
+                },
+            );
+            return pt.intern(.{ .aggregate = .{
+                .ty = ty,
+                .storage = .{ .elems = &.{} },
+            } });
+        },
+        .array_literal => |nodes| {
+            const types = try self.sema.arena.alloc(InternPool.Index, nodes.len);
+            const values = try self.sema.arena.alloc(InternPool.Index, nodes.len);
+            for (0..nodes.len) |i| {
+                values[i] = try self.lowerExprAnonResTy(nodes.at(@intCast(i)));
+                types[i] = Value.fromInterned(values[i]).typeOf(pt.zcu).toIntern();
+            }
+            const ty = try ip.getTupleType(
+                gpa,
+                pt.tid,
+                .{
+                    .types = types,
+                    .values = values,
+                },
+            );
+            return pt.intern(.{ .aggregate = .{
+                .ty = ty,
+                .storage = .{ .elems = values },
+            } });
+        },
+        .struct_literal => |init| {
+            const elems = try self.sema.arena.alloc(InternPool.Index, init.names.len);
+            for (0..init.names.len) |i| {
+                elems[i] = try self.lowerExprAnonResTy(init.vals.at(@intCast(i)));
+            }
+            const struct_ty = switch (try ip.getStructType(
+                gpa,
+                pt.tid,
+                .{
+                    .layout = .auto,
+                    .fields_len = @intCast(init.names.len),
+                    .known_non_opv = false,
+                    .requires_comptime = .no,
+                    .any_comptime_fields = true,
+                    .any_default_inits = true,
+                    .inits_resolved = true,
+                    .any_aligned_fields = false,
+                    .key = .{ .reified = .{
+                        .zir_index = self.base_node_inst,
+                        .type_hash = @intFromEnum(node),
```

Comment:
This almost works great, but unfortunately it doesn't quite play nice with incremental; with your current code, changing the number of fields in a ZON struct on an incremental update would crash the compiler. We could fix it with some hacky incremental dependency stuff, but it would be nicer to instead change this to incorporate `elems` into the hash; the idea behind `reified` is that the `type_hash` should include the type's "identity" as well as the source location. So, you'll want to do something like this (see similar logic in `Sema.structInitAnon`:
```suggestion
                        .type_hash = hash: {
                            var hasher: std.hash.Wyhash = .init(0);
                            hasher.update(std.mem.asBytes(node));
                            hasher.update(std.mem.sliceAsBytes(elems));
                            break :hash hasher.final();
                        },
```

Reply:
Ah that makes sense. I added the following to the hash:
* The node
* The zoir null terminated strings for the field names
* The interned types, and interned values

Is this correct, or if a field changed its name without changing its length would there be a collision?